STACpopulator/models.py

```diff
@@ -0,0 +1,87 @@
+import datetime as dt
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import (
+    AnyHttpUrl,
+    AnyUrl,
+    BaseModel,
+    Field,
+    SerializeAsAny,
+    field_validator,
+)
+from typing_extensions import TypedDict
+
+
+class Geometry(TypedDict):
+    type: str
+    coordinates: List[List[List[float]]]
+
+
+class Asset(BaseModel):
+    href: AnyHttpUrl
+    media_type: Optional[str] = None
+    title: Optional[str] = None
+    description: Optional[str] = None
+    roles: Optional[List[str]] = None
+
+
+class STACItemProperties(BaseModel):
+    """Base STAC Item properties data model. In concrete implementations, users would want to define a new
+    data model that inherits from this base model and extends it with properties tailored to the data they are
+    ingesting."""
+
+    start_datetime: Optional[dt.datetime] = None
+    end_datetime: Optional[dt.datetime] = None
+    datetime: Optional[dt.datetime] = None
+
+    @field_validator("datetime", mode="before")
+    @classmethod
+    def validate_datetime(cls, v: Union[dt.datetime, str], values: Dict[str, Any]) -> dt:
+        if v == "null":
+            if not values["start_datetime"] and not values["end_datetime"]:
+                raise ValueError("start_datetime and end_datetime must be specified when datetime is null")
+
+
+# class Link(BaseModel):
+#     """
+#     https://github.com/radiantearth/stac-spec/blob/v1.0.0/collection-spec/collection-spec.md#link-object
+#     """
+
+#     href: str = Field(..., alias="href", min_length=1)
+#     rel: str = Field(..., alias="rel", min_length=1)
+#     type: Optional[str] = None
+#     title: Optional[str] = None
+#     # Label extension
+#     label: Optional[str] = Field(None, alias="label:assets")
+#     model_config = ConfigDict(use_enum_values=True)
+
+#     def resolve(self, base_url: str) -> None:
+#         """resolve a link to the given base URL"""
+#         self.href = urljoin(base_url, self.href)
+
+
+# class PaginationLink(Link):
+#     """
+#     https://github.com/radiantearth/stac-api-spec/blob/master/api-spec.md#paging-extension
+#     """
+
+#     rel: Literal["next", "previous"]
+#     method: Literal["GET", "POST"]
+#     body: Optional[Dict[Any, Any]] = None
+#     merge: bool = False
+
+
+# Links = RootModel[List[Union[PaginationLink, Link]]]
```

Comment:
Not sure if this gets lost after conversion? 

At the very least, the converted STAC Item should contain the `rel: source` with `href` point at the original NetCDF file and the `rel: describes` with `href` pointing at the NCML. In the case of `rel: source`, it is also very important that `title` contains the "path" of the file without the THREDDS URL prefix and the "service" path element (dodsC, fileServer, etc.).

Reply:
When the STAC Item is created (POST'd to the API), some resources in Magpie are generated under the `stac` service to manage its permissions.
Since that item refers to a NetCDF under THREDDS, a corresponding resource exists under `thredds` service to manage its permissions as well.
Using https://github.com/bird-house/birdhouse-deploy/pull/386, the intent is that when a permission is updated on one, it will synchronize with the other, such that viewing the NetCDF metadata from either service is allowed/denied in the same way for a given user or group.

The problem in doing the above is that the URLs of these 2 services are not "compatible". 
The generated resources would be nested like so:
```yaml
stac
  collections
    <directory-uuiid>
      items
        <another-uuid>
thredds
  <directory-name>
    <netcdf-name>
```
Since one works with UUIDs and the other with path names, there is no direct way to guess the references only from the URLs. Therefore, the STAC hook will use one of the references in `links`, namely the `rel: source` one, to obtain the STAC Asset URL of the original THREDDS NetCDF location. To establish the "link" between the `stac` and `thredds` Magpie services, it will use the `title` value from the `rel: source` link to establish a `resource_display_name` in Magpie. That parameter will be readable from Cowbird later on to resolve the `stac` UUID <-> `thredds` name resource relationship. 

For example, the following catalog reference: 
https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/catalog/birdhouse/testdata/xclim/cmip6/catalog.html?dataset=birdhouse/testdata/xclim/cmip6/sic_SImon_CCCma-CanESM5_ssp245_r13i1p2f1_2020.nc
Should have (at least) the following link in the STAC Item
```json
{
  "rel": "source",
  "title": "birdhouse/testdata/xclim/cmip6/sic_SImon_CCCma-CanESM5_ssp245_r13i1p2f1_2020.nc",
  "href": "https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/fileServer/birdhouse/testdata/xclim/cmip6/sic_SImon_CCCma-CanESM5_ssp245_r13i1p2f1_2020.nc",
  "type": "application/x-netcdf"
}
```

Reply:
Thanks for that explanation @fmigneault. This makes sense. Let me implement this.

Reply:
Okay done. If all looks good then I will complete the merge.